design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,272 @@
+
+## Implementation Phases
+
+**Phase1** - CDI APIs
+* VolumeReference API/controller
+* VolumeImportPoller API/controller
+
+**Phase2** - SSP + Common Templates
+* Common Templates transition to VolumeReferences for DataVolumeTemplate source
+* SSP Operator creates VolumeImportPollers for automatic PVC import
+
+**Phase3** - HCO and UI
+* HCO Team integrates with SSP Operator’s API for setting volume import sources
+* UI Team transitions to detecting boot source using a linked VolumeReference rather than looking directly for a PVC of a specific name.
+
+# API Examples
+## VM + DataVolumeTemplate + VolumeReference as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+  pvc:
+    accessModes:
+    - ReadWriteOnce
+    resources:
+      requests:
+        storage: 2Gi
+  source:
+    volumeReference:
+      name: fedora
+      namespace: golden-images 
+```
+
+## VolumeImportPoller Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the VolumeImportPoller will import them as new PVCs and automatically manage updating the corresponding VolumeReference.
+
+* **source** specifies where to poll from
+* **managedVolumeReference** specifies the name of the corresponding VolumeReference this poller will manage. The controller should create this VolumeReference if it doesn't exist, and adopt ownership of the VolumeReference if it does already exist.
+* **pollInterval** specifies how often to look for new imports 
+* **garbageCollectOutdated** specifies whether old PVCs should be cleaned up after a new PVC is imported.
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportPoller
+metadata:
+  name: fedora-poller
+  namespace: golden-images
+spec:
+  source:
+    registry:
+      url: docker://quay.io/kubevirt/fedora
+  pollInterval: 12h
+  garbageCollectOutdated: true
+  managedVolumeReference: fedora
+status:
+  lastImportedPVC: quay.io/kubevirt/fedora:33
+```
+
+## VolumeReference Example
+
+This example shows how a PVC is linked using the claimName field on the VolumeReference spec. The VolumeReference’s status section will indicate if the claimName is linked or dangling. 
+
+The UI can use the phase field to determine if a boot source exists for a VolumeReference or not. 
+
+* **Phase: linked** indicates that the the claimName is set and that the corresponding PVC exists and is populated
+* **Phase: dangling** indicates that the claimName points to a non-existent PVC, meaning the underlying PVC has either been deleted or is not created yet. 
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeReference
+metadata:
+  name: fedora
+spec:
+  source:
+    pvc:
+      namespace: golden-images 
+      name: fedora-32-001
+status:
+  phase: linked
```

Comment:
If there are only two possible values, maybe a bool like `readyToUse` in `VolumeSnapshots`?

Reply:
> If there are only two possible values, maybe a bool like readyToUse in VolumeSnapshots?

maybe we should avoid phase entirely, and use simply a `Ready` condition instead. That allows us to assign a reason to why the VolumeReference isn't bound... like `PVC doesn't exist`, or perhaps in the future `http endpoint inaccessible` or `unauthorized access`

wdyt?

Reply:
Condition is good